lib/phoenix_live_view/channel.ex

```diff
@@ -900,16 +946,102 @@ defmodule Phoenix.LiveView.Channel do
     Phoenix.LiveView.assign(socket, :live_action, action)
   end
 
-  defp maybe_update_uploads(state, %{"uploads" => uploads}) do
-    Enum.reduce(uploads, state, fn {ref, entries}, acc ->
-      upload_conf = Upload.get_upload_by_ref!(acc.socket, ref)
+  defp maybe_update_uploads(%Socket{} = socket, %{"uploads" => uploads} = payload) do
+    cid = payload["cid"]
+
+    Enum.reduce(uploads, socket, fn {ref, entries}, acc ->
+      upload_conf = Upload.get_upload_by_ref!(acc, ref)
 
-      case Upload.put_entries(acc.socket, upload_conf, entries) do
-        {:ok, new_socket} -> %{acc | socket: new_socket}
-        {:error, _error_resp, %Socket{} = new_socket} -> %{acc | socket: new_socket}
+      case Upload.put_entries(acc, upload_conf, entries, cid) do
+        {:ok, new_socket} -> new_socket
+        {:error, _error_resp, %Socket{} = new_socket} -> new_socket
       end
     end)
   end
 
-  defp maybe_update_uploads(state, %{} = _payload), do: state
+  defp maybe_update_uploads(%Socket{} = socket, %{} = _payload), do: socket
+
+  defp register_entry_upload(state, from, info) do
+    %{channel_pid: pid, ref: ref, entry_ref: entry_ref, cid: cid} = info
+
+    {new_state, return} =
+      with_socket(state, cid, nil, fn socket, _ ->
+        conf = Upload.get_upload_by_ref!(socket, ref)
+
+        case Upload.register_entry_upload(socket, conf, pid, entry_ref) do
+          {:ok, new_socket, entry} ->
+            reply = %{max_file_size: entry.client_size, chunk_timeout: conf.chunk_timeout}
+            GenServer.reply(from, {:ok, reply})
+            {new_socket, {:ok, nil}}
+
+          {:error, reason} ->
+            GenServer.reply(from, {:error, reason})
+            {socket, :error}
+        end
+      end)
+
+    case return do
+      {:ok, _} -> put_upload_pid(new_state, pid, ref, entry_ref, cid)
+      :error -> {:noreply, new_state}
+    end
+  end
+
+  # If :error is returned, the socket must not change,
+  # otherwise we need to call push_diff on all cases.
+  defp with_socket(state, nil, ref, fun) do
+    {new_socket, return} = fun.(state.socket, nil)
+
+    case return do
+      {:ok, ref_reply} ->
+        {:noreply, new_state} = handle_changed(state, new_socket, ref_reply)
+        {new_state, return}
+
+      {:ok, ref_reply, extra} ->
+        {:noreply, new_state} = handle_changed(state, new_socket, ref_reply)
+        {new_state, extra}
+
+      :error ->
+        {push_noop(state, ref), return}
+    end
+  end
+
+  defp with_socket(state, cid, ref, fun) do
+    %{socket: socket, components: components} = state
+
+    {new_state, diff, return} =
+      case Diff.with_component(socket, cid, %{}, components, fun) do
+        {diff, new_components, return} ->
+          {%{state | components: new_components}, diff, return}
+
+        :error ->
+          {state, :error}
```

Comment:
This should be a three element tuple but note my proposed refactoring above fixes this.